keps/4400-workload-priority-preemption-policy/README.md

```diff
@@ -0,0 +1,141 @@
+# KEP-4400: WorkloadPriorityClass Preemption Policy
+
+<!-- toc -->
+- [Summary](#summary)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [Story 1](#story-1)
+  - [Notes/Constraints/Caveats (Optional)](#notesconstraintscaveats-optional)
+    - [Quota Constraints for Non-Preemptible Workloads](#quota-constraints-for-non-preemptible-workloads)
+- [Design Details](#design-details)
+  - [API Changes](#api-changes)
+  - [Implementation Details](#implementation-details)
+  - [Interaction with ClusterQueue Preemption Policies](#interaction-with-clusterqueue-preemption-policies)
+  - [Test Plan](#test-plan)
+    - [Unit Tests](#unit-tests)
+    - [Integration tests](#integration-tests)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+This KEP proposes extending the `WorkloadPriorityClass` API (introduced in KEP-973) with a `preemptionPolicy` field to provide administrators with fine-grained control over which workloads can be preempted. This allows multiple workloads within the same ClusterQueue to have different preemption behaviors irrespective of priority level.
+
+## Proposal
+
+Extend the `WorkloadPriorityClass` API with a `preemptionPolicy` field that controls whether workloads using that priority class can be preempted. This follows the same pattern as Kubernetes core `PriorityClass.preemptionPolicy`.
+
+The preemption policy will be enforced at scheduling time, with additional quota constraints for non-preemptible workloads to prevent resource deadlocks.
+
+### User Stories
+
+#### Story 1
+
+As a Kueue user, I want to make specific workloads non-preemptible because they cannot be resumed easily, and I'm willing to wait longer for these workloads to run. My administrator will use the PreemptionPolicy field in WorkloadPriorityClass to provide a lever for this behavior.
+
+### Notes/Constraints/Caveats (Optional)
+
+#### Quota Constraints for Non-Preemptible Workloads
+
+Workloads using a `WorkloadPriorityClass` with `preemptionPolicy: Never` are subject to the additional quota constraint:
+
+- **Nominal Quota Limit**: They can only consume resources up to the ClusterQueue's nominal quota
+- **No Borrowing**: They cannot borrow resources from other ClusterQueues in the cohort  
+
+This prevents scenarios where non-preemptible workloads consume borrowed quota that cannot be reclaimed
+
+## Design Details
+
+### API Changes
+
+Extend the `WorkloadPriorityClass` type with a `preemptionPolicy` field:
+
+```golang
```

Comment:
Could you remove an existing fields from this code snippet, then leave only proposed fields?